interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/TypeInfo.scala

```diff
@@ -0,0 +1,153 @@
+package pl.touk.nussknacker.engine.definition
+
+import cats.data.ValidatedNel
+import cats.implicits.catsSyntaxValidatedId
+import io.circe.Encoder
+import io.circe.generic.JsonCodec
+import pl.touk.nussknacker.engine.api.generics.{NoVarArgumentTypeError, SpelParseError, VarArgumentTypeError}
+import pl.touk.nussknacker.engine.api.typed.TypeEncoders
+import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypedClass, TypingResult}
+
+object TypeInfo {
+  //a bit sad that it isn't derived automatically, but...
+  private implicit val tce: Encoder[TypedClass] = TypeEncoders.typingResultEncoder.contramap[TypedClass](identity)
+
+  @JsonCodec(encodeOnly = true) case class Parameter(name: String, refClazz: TypingResult)
+
+  @JsonCodec(encodeOnly = true) case class SerializableMethodInfo(parameters: List[Parameter],
+                                                                  refClazz: TypingResult,
+                                                                  description: Option[String],
+                                                                  varArgs: Boolean)
+
+  implicit val methodInfoEncoder: Encoder[MethodInfo] = Encoder[SerializableMethodInfo].contramap(_.serializable)
+
+  object MethodInfo {
+    private val arrayClass = classOf[Array[Object]]
+
+    def apply(parameters: List[Parameter],
+              refClazz: TypingResult,
+              name: String,
+              description: Option[String],
+              varArgs: Boolean): StaticMethodInfo = (varArgs, parameters) match {
+      case (true, noVarArgParameters :+ Parameter(paramName, TypedClass(`arrayClass`, varArgType :: Nil))) =>
+        VarArgsMethodInfo(noVarArgParameters, Parameter(paramName, varArgType), refClazz, name, description)
+      case (true, _ :+ Parameter(_, TypedClass(`arrayClass`, _))) =>
+        throw new AssertionError("Array must have one type parameter")
+      case (true, _ :+ Parameter(_, _)) =>
+        throw new AssertionError("VarArg must have type of array")
+      case (true, Nil) =>
+        throw new AssertionError("Method with varArgs must have at least one parameter")
+      case (false, _) =>
+        NoVarArgsMethodInfo(parameters, refClazz, name, description)
+    }
+  }
+
+  sealed trait MethodInfo {
+    def apply(arguments: List[TypingResult]): ValidatedNel[SpelParseError, TypingResult]
+
+    def name: String
+
+    def staticParameters: List[Parameter]
+
+    def staticResult: TypingResult
+
+    def description: Option[String]
+
+    def varArgs: Boolean
+
+    def serializable: SerializableMethodInfo =
+      SerializableMethodInfo(staticParameters, staticResult, description, varArgs)
+  }
+
+  sealed trait StaticMethodInfo extends MethodInfo {
+    protected def checkNoVarArguments(arguments: List[TypingResult], parameters: List[Parameter]): Boolean =
+      arguments.length == parameters.length &&
+        arguments.zip(parameters).forall{ case(arg, param) => arg.canBeSubclassOf(param.refClazz)}
```

Comment:
why we consider different types as var args problem?

Reply:
I don't understand the question

Reply:
I didn't understand the code, sorry ;)